client/state/data-layer/test/extensions-middleware.js

```diff
@@ -0,0 +1,239 @@
+import { local } from '../utils';
+
+describe( 'Calypso Extensions Data Layer Middleware', () => {
+	let next;
+	let store;
+
+	beforeEach( () => {
+		next = spy();
+
+		store = {
+			dispatch: spy(),
+			getState: stub(),
+			replaceReducers: spy(),
+			subscribe: spy(),
+		};
+
+		store.getState.returns( Object.create( null ) );
+	} );
+
+	it( 'should pass along actions without corresponding handlers', () => {
+		const action = { type: 'UNSUPPORTED_ACTION' };
+
+		const config = configureMiddleware( Object.create( null ), Object.create( null ) );
+		addHandlers( 'my-extension', {}, config );
+
+		config.middleware( store )( next )( action );
+
+		expect( store.dispatch ).to.not.have.beenCalled;
+		expect( next ).to.have.been.calledWith( action );
+	} );
+
+	it( 'should pass along local actions untouched', () => {
+		const adder = spy();
+		const handlers = {
+			[ 'ADD' ]: [ adder ],
+		};
+
+		const config = configureMiddleware( Object.create( null ), Object.create( null ) );
+		addHandlers( 'my-extension', handlers, config );
+		const action = local( { type: 'ADD' } );
+
+		config.middleware( store )( next )( action );
+
+		expect( next ).to.have.been.calledWith( action );
+		expect( adder ).to.not.have.beenCalled;
+	} );
+
+	it( 'should not pass along non-local actions with non data-layer meta', () => {
+		const adder = spy();
+		const handlers = {
+			[ 'ADD' ]: [ adder ],
+		};
+
+		const config = configureMiddleware( Object.create( null ), Object.create( null ) );
+		addHandlers( 'my-extension', handlers, config );
+		const action = { type: 'ADD', meta: { semigroup: true } };
+
+		config.middleware( store )( next )( action );
+
+		expect( next ).to.not.have.beenCalled;
+		expect( adder ).to.have.been.calledWith( store, action );
+	} );
+
+	it( 'should not pass along non-local actions with data-layer meta but no bypass', () => {
+		const adder = spy();
+		const handlers = {
+			[ 'ADD' ]: [ adder ],
+		};
+
+		const config = configureMiddleware( Object.create( null ), Object.create( null ) );
+		addHandlers( 'my-extension', handlers, config );
+		const action = { type: 'ADD', meta: { dataLayer: { data: 42 } } };
+
+		config.middleware( store )( next )( action );
+
+		expect( next ).to.not.have.beenCalled;
+		expect( adder ).to.have.been.calledWith( store, action );
+	} );
+
+	it( 'should intercept actions in appropriate handler', () => {
+		const adder = spy();
+
+		const handlers = {
+			[ 'ADD' ]: [ adder ],
+		};
+
+		const config = configureMiddleware( Object.create( null ), Object.create( null ) );
+		addHandlers( 'my-extension', handlers, config );
+		const action = { type: 'ADD' };
+
+		config.middleware( store )( next )( action );
+
+		expect( next ).to.not.have.been.calledWith( action );
+		expect( adder ).to.have.been.calledWith( store, action );
+	} );
+
+	it( 'should allow continuing the action down the chain', () => {
+		const adder = spy( ( _store, _action, _next ) => _next( _action ) );
```

Comment:
I'll admit, I shamelessly grabbed this test from `test/wpcom-api-middleware.js`

However, it's verifying that `next()` was called with `local( action )`, which is what I figure is important here. It needs to get action from its own parameter list instead of just relying on the JS scope in which it was created to get it, in order to be a valid test.

There are many other tests which verify that `adder()` gets called correctly, so that's not the primary focus for this test.